### ISSUE TITLE
bug in MOTOR loss

### ISSUE THREAD

**Issue by duncanmcelfresh** [som-shahlab/femr#213]
it looks like there is a bug in the MOTOR loss in (src/femr/models/transformer.py)

Based on equation 1 from the paper, the MOTOR log-loss (actually the log-likelihood) is:

 \delta_{ijkp} * ( -\lambda_{ijkp} *U_{ijkp} + Log(\lambda_{ijkp}) ) + (1 - \delta_{ijkp}) * ( -\lambda_{ijkp} * U_{ijkp} )

Which simplifies to…

\delta_{ijkp} * Log(\lambda_{ijkp}) - ( \lambda_{ijkp} * U_{ijkp} )

Since above is the log-likelihood, we want to maximize this, or minimize the negative of this -- in other words we want to minimize the following loss:

( \lambda_{ijkp} * U_{ijkp} ) - \delta_{ijkp} * Log(\lambda_{ijkp})

To implement this, we make the following replacements (using the same names as in the femr code):

`log_time` <== Log2(U_{ijkp})
`time_dependent_logits` <== Log(\lambda_{ijkp})
`is_event` <== \delta_{ijkp}

Now the loss is:

[A]  `( exp( time_dependent_logits ) * exp2( log_time ) ) - is_event * time_dependent_logits`
 
And using this identity: b^x = a^(log_a(b) *x)

We replace the first term with exp2( Log(2) * time_dependent_logits ) … and the loss is:

[B] `exp2( Log(2) * time_dependent_logits ) * exp2( log_time ) - is_event * time_dependent_logits`

combining the two exponentials, 

[C] `exp2( ( Log(2) * time_dependent_logits ) + log_time) - is_event * time_dependent_logits`

And this is different from the implementation I see in transformer.py:

[femr] `exp2( time_dependent_logits + log_time ) - Log(2) * is_event * time_dependent_logits`

(the Log(2) is in a different spot).

Is this wrong? My guess is that -- during implementation -- the Log(2) was accidentally pulled out of the exponential in [B] and treated as a factor on the first term.

**Comment by EthanSteinberg** [comment]
I don't think this is a bug in the transformer.py. I think your formula is just an alternate way of parameterizing lambda for the same loss.

In your math, you have defined lambda = e^(time_dependent_logits)

In my math I have defined lambda = 2^(time_dependent_logits).

Both should converge to the same result, but with time_dependent_logits that are a constant factor (log(2)) away from each other.

Similarly, you could also define lambda = 3^(time_dependent_logits) and things would still work.

What is important is that the survival loss and the event loss are in sync.

(Alternative we could also consider linear risk, aka lambda = time_dependent_logits, which would result in different formulas).

**Comment by EthanSteinberg** [comment]
In case you are interested, the reason I constructed it the way I did was because your formula has more floating point multiplications than mine, as it requires a log2 factor broadcasted across an array, while mine only has a log2 factor multiplied by a scalar.

**Comment by duncanmcelfresh** [comment]
yup agreed this is fine
